lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// +build !solaris,!darwin solaris,cgo darwin,cgo
+
+package fswatcher
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/ignore"
+	"github.com/zillode/notify"
+)
+
+type eventType int
+
+const (
+	nonRemove eventType = 1
+	remove              = 2
+	mixed               = 3
+)
+
+// Not meant to be changed, but must be changeable for tests
+var (
+	maxFiles       = 512
+	maxFilesPerDir = 128
+)
+
+type event struct {
+	path         string
+	firstModTime time.Time
+	lastModTime  time.Time
+	evType       eventType
+}
+
+type eventBatch map[string]*event
+
+type eventDir struct {
+	path      string
+	parentDir *eventDir
+	events    eventBatch
+	dirs      map[string]*eventDir
```

Comment:
Yes (probably). Dirs only contains those dirs that (recursively) contain events themselves.
The struct got smaller due to the path optimizations (thanks) and there are comments explaining what the structs `event` and `eventDir` are - I hope those are clear?
I am currently doubting that the term "event" is appropriate here, as it might track multiple events to a specific file - maybe that also contributes to this not being overly clear.